phantazm-commons/src/main/java/com/github/phantazmnetwork/commons/vector/Region3I.java

```diff
@@ -0,0 +1,110 @@
+        Vec3I second = Vec3I.of(first.getX() + lengths.getX() - 1, first.getY() + lengths.getY() - 1, first
+                .getZ() + lengths.getZ() - 1);
+        return encompassing(first, second, Vec3I.ORIGIN);
+    }
+
+    static boolean overlaps(int oX1, int oY1, int oZ1, int lX1, int lY1, int lZ1,
+                            int oX2, int oY2, int oZ2, int lX2, int lY2, int lZ2) {
+        int nx1 = oX1 + lX1;
+        int ny1 = oY1 + lY1;
+        int nz1 = oZ1 + lZ1;
+
+        int nx2 = oX2 + lX2;
+        int ny2 = oY2 + lY2;
+        int nz2 = oZ2 + lZ2;
+
+        return Math.min(oX1, nx1) < Math.max(oX2, nx2) && Math.max(oX1, nx1) > Math.min(oX2, nx2)
+                && Math.min(oY1, ny1) < Math.max(oY2, ny2) && Math.max(oY1, ny1) > Math.min(oY2, ny2)
+                && Math.min(oZ1, nz1) < Math.max(oZ2, nz2) && Math.max(oZ1, nz1) > Math.min(oZ2, nz2);
+    }
+
+    static boolean overlaps(@NotNull Vec3I firstOrigin, @NotNull Vec3I firstLengths,
+                            @NotNull Vec3I secondOrigin, @NotNull Vec3I secondLengths) {
+        return overlaps(firstOrigin.getX(), firstOrigin.getY(), firstOrigin.getZ(), firstLengths.getX(), firstLengths
+                .getY(), firstLengths.getZ(), secondOrigin.getX(), secondOrigin.getY(), secondOrigin.getZ(),
+                secondLengths.getX(), secondLengths.getY(), secondLengths.getZ());
+    }
+
+    static boolean overlaps(@NotNull Region3I first, @NotNull Region3I second) {
+        return overlaps(first.getOrigin(), first.getLengths(), second.getOrigin(), second.getLengths());
+    }
+
+    default int volume() {
+        Vec3I lengths = getLengths();
+        return lengths.getX() * lengths.getY() * lengths.getZ();
+    }
+
+    default boolean overlaps(@NotNull Region3I other) {
+        return overlaps(this, other);
+    }
+
+    @Override
+    default @NotNull Iterator<Vec3I> iterator() {
+        return new Iterator<>() {
+            private final Vec3I origin = Region3I.this.getOrigin();
+            private final Vec3I lengths = Region3I.this.getLengths();
+            private int x = origin.getX();
+            private int y = origin.getY();
+            private int z = origin.getZ();
+
+            @Override
+            public boolean hasNext() {
+                return z < (origin.getZ() + lengths.getZ());
+            }
+
+            @Override
+            public Vec3I next() {
+                int curX = x;
+                int curY = y;
+                int curZ = z;
+
+                if(++x >= origin.getX() + lengths.getX()) {
+                    curX = x = origin.getX();
+                    if(++y >= (origin.getY() + lengths.getY())) {
+                        curY = y = origin.getY();
+                        if(++z > (origin.getZ() + lengths.getZ())) {
+                            throw new NoSuchElementException();
+                        }
+                    }
+                }
```

Comment:
iterating across a diagonal?

Reply:
no, y is only iterated when x >= xEnd, z is only iterated when y >= yEnd